File: javaee/persistence/eclipselink/src/main/java/org/sgodden/tom/domain/CustomerOrderRepositoryImpl.java
package org.sgodden.tom.domain;

import javax.persistence.EntityManager;
import javax.persistence.Query;

import org.sgodden.transaction.Transactional;

import com.google.inject.Inject;
import com.google.inject.Provider;
import com.google.inject.Singleton;

@Singleton
public class CustomerOrderRepositoryImpl implements CustomerOrderRepository {

	@Inject
	private Provider<EntityManager> em;

	@Transactional
	public void persist(CustomerOrder order) {
		em.get().persist(order);
	}

	@Transactional
	public CustomerOrder findById(long id) {
		return em.get().find(CustomerOrder.class, id);
	}

	public long count() {
		Query q = em.get().createQuery("select count(o) from CustomerOrder o");
		return (Long) q.getSingleResult();
	}

}

File: templates/scala-mongodb/persistence/src/main/scala/org/sgodden/tom/persistence/CustomerOrderRepositoryImpl.scala
package org.sgodden.tom.persistence

import com.novus.salat._
import com.novus.salat.global._
import com.mongodb.casbah.Imports._
import com.mongodb.casbah.commons.conversions.scala._
import model.CustomerOrderP
import org.sgodden.tom.model.{CustomerOrder, ICustomerOrder, CustomerOrderRepository}

class CustomerOrderRepositoryImpl(databaseName: String) extends CustomerOrderRepository {
  val conn = MongoConnection()
  val db = conn(databaseName)
  val coll = db("customerOrders")

  println("ENV: " + System.getProperty("env"))

  RegisterConversionHelpers()
  RegisterJodaTimeConversionHelpers()

  def remove(order: ICustomerOrder) {
    order.asInstanceOf[CustomerOrder].approveRemove(() => {
      coll.remove(MongoDBObject("_id" -> new ObjectId(order.getId)))
    })
  }

  def findAll = {
    coll.map(dbo => {
      grater[CustomerOrderP].asObject(dbo).asObject
    }).toList
  }

  def persist(order: ICustomerOrder) {
    order.asInstanceOf[CustomerOrder].approvePersist(() => {
      coll.save(grater[CustomerOrderP].asDBObject(CustomerOrderP(order)))
      order.asInstanceOf[CustomerOrder].setId(coll.last.get("_id").toString)
    })
  }

  def merge(order: ICustomerOrder) {
    order.asInstanceOf[CustomerOrder].approvePersist(() => {
      coll.save(grater[CustomerOrderP].asDBObject(CustomerOrderP(order)))
    })
  }

  def findById(id: String) =
    grater[CustomerOrderP].asObject(coll.findOne(MongoDBObject("_id" -> new ObjectId(id))).get).asObject

  def count = {
    coll.size
  }

}
